File: GFG_Java/LinkedLists/findMiddle.java
public class findMiddle {
	class Node {
		int data;
		Node next;
		Node(int d) {data = d; next = null;}
	}

	Node head;

	public void push(int data) {
		Node newNode = new Node(data);
		if(head == null) {
			head = newNode;
			return;
		}
		newNode.next = head;
		head = newNode;
	}

	public void printList() {
		Node temp = head;
		while(temp != null) {
			System.out.print(temp.data + " ");
			temp = temp.next;
		}
		System.out.println();
	}

	public void printMiddle() {
		Node mid = head;
		int count = 0;
		while(head != null) {
			if((count % 2) == 1)
				mid = mid.next;
			++count;
			head = head.next;
		}
		System.out.println(mid.data);
	}

	public static void main(String args[]) {
		findMiddle llist = new findMiddle();
		llist.push(17);
		llist.push(16);
		llist.push(15);
		llist.push(14);
		llist.push(13);
		llist.push(12);
		llist.push(11);
		llist.printList();
		System.out.print("The middle node is: ");
		llist.printMiddle();
	}
}

File: GFG_C/Additional_Exercises/SinglyLinkedLists/findMiddle.c
#include <stdio.h>
#include <stdlib.h>

#define SUCCESS 0

typedef int data_t;
typedef struct _node {
    data_t data;
    struct _node *next;
}Node;

int createALinkedList(Node **head);
int printList(Node *head);
int printMiddle(Node *head);

int main() {
    // create a linked list
    Node *head = NULL;
    createALinkedList(&head);
    printList(head);
    
    printf("middle node: %d\n", printMiddle(head));
    return SUCCESS;
}

int printList(Node *head) {
    while(head) {
        printf("%d->", head->data);
        head = head->next;
    }
    puts("NULL");
    putc('\n', stdout);
}

int createALinkedList(Node **head) {
    int i = 0;
    (*head) = (Node *)malloc(sizeof(Node));
    Node *temp = *head;
    while(i < 9) {
        (temp)->data = i++;
        (temp)->next = (Node *)malloc(sizeof(Node));
        temp = temp->next;
    }
    temp->data = i;
    temp->next = NULL;

    return SUCCESS;
}

int printMiddle(Node *head) {
    Node *slow = head;
    Node *fast = head;

    while(fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
    }

    return slow->data;
}